src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,155 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<>(tail, value, null);
+        if (tail == null) {
+            head = node;
+        } else {
+            tail.next = node;
+        }
+        tail = node;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Index " + index + " out of bounds.");
+        }
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> nodeByIndex = getNodeByIndex(index);
+            Node<T> prevNodeByIndex = nodeByIndex.prev;
+            Node<T> newNode = new Node<>(prevNodeByIndex, value, nodeByIndex);
+            nodeByIndex.prev = newNode;
+            if (prevNodeByIndex == null) {
+                head = newNode;
+            } else {
+                prevNodeByIndex.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> neededNode = head;
+        for (int i = 0; i < index; i++) {
+            neededNode = neededNode.next;
+        }
+        return neededNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> setNode = getNodeByIndex(index);
```

Comment:
```suggestion
        Node<T> setNode = getNodeByIndex(index);
```
we have checkIndex() method in getNodeByIndex(), so we can remove it from here